docs/framework/architecture/ui-library.md

```diff
@@ -439,6 +441,170 @@ dropdownView.bind( 'isEnabled' ).toMany( buttons, 'isEnabled',
+
+Dialog is a popup window that does not close when the user clicks outside of it. It allows for interacting with the editor and its content while being open (unless it is a modal, which blocks the interaction with the rest of the page until it is closed). When configured to display a [header](#header), a dialog is also {@link module:ui/bindings/draggableviewmixin~DraggableViewMixin draggable} using mouse or touch. Only one dialog can be open at a time - opening another one closes the previously visible one.
+
+A minimal plugin creating a toolbar button component showing an empty dialog can look like this:
+
+```js
+import { Dialog, ButtonView } from '@ckeditor/ckeditor5-ui';
+
+class MyPluginWithDialog extends Plugin {
+	public static get requires() {
+		return [ Dialog ] as const;
+	}
+
+	public init(): void {
+		this.editor.ui.componentFactory.add( 'aiAssistant', locale => {
+			const button = new ButtonView( locale );
+
+			button.set( {
+				label: 'Show a dialog'
+			} );
+
+			this.listenTo( button, 'execute', () => {
+				const dialog = this.editor.plugins.get( 'Dialog' );
+
+				dialog.show( {
+					// The dialog definition can be defined here.
+				} );
+			} );
+
+			return button;
+		} );
+	}
+}
+```
+
+Modals are similar to the dialogs - they share the same structure rules, API etc. The major difference is that while a modal is open, the user can not interact with the editor or the rest of the page - it is covered with a non-transparent overlay. It can be used e.g. to enforce user to take one of the specified actions.
+
+To create a modal, use the optional {@link module:ui/dialog/dialog~DialogDefinition#isModal `isModal`} argument of the {@link module:ui/dialog/dialog~Dialog#show `Dialog#show()`} method:
+
+```js
+dialog.show( {
+	isModal: true,
+
+	// The rest of the dialog definition...
+} );
+```
+
+There are different ways to close a modal: either by clicking the "Close" button in the corner (if the [header](#header) is visible), an <kbd>Esc</kbd> key press or one of the [action buttons](#action-buttons).
+
+#### Structure and behaviour
+
+A dialog can have three parts, each of which is optional: the [header](#header) (also used as a drag handler), the [content](#contents) (the body of the dialog) and the [action buttons](#action-buttons) area (a collection of buttons). The order of the parts can not be changed.
+
+##### Header
+
+A header may consist of any combination of three elements:
+
+* the icon,
+* the title,
+* the "Close" button.
+
+The "Close" button is added automatically as long as you provide an icon or a title. The code below shows how to create display a dialog with a header that comes with an icon and a title, but without the "Close" button.
```

Comment:
```suggestion
By default, the "Close" button ("X") is added to the header as long as you provide an icon or a title. To hide it, set `hasCloseButton` flag to `false`:
```

Reply:
Do we want to add:

*   some kind of info that if you hide the close button you should add a way to close the dialog (kind of obvious, I know),
*   somewhere information that the dialog can be closed using `Esc` key (and whether it still true if `hasCloseButton` is false),
    *   okay I see we have the info about `Esc` key in a11y section.
*   info how to prevent closing on `Esc`?
    *   or how to prevent closing in general (especiall in regards to modals)
    *   of info that this is a bad practice
*   info if model can be closed using `Esc`?
    *   this and above points are related